lib/cpp/src/thrift/transport/TSocketUtils.h

```diff
@@ -0,0 +1,185 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ * Copyright (c) 2020 Max Ivanov <ulidtko@gmail.com>
+ */
```

Comment:
Please move any extra info (line 21) to NOTICE
https://www.apache.org/dev/licensing-howto.html

Reply:
Hey Jens, sorry I don't get this... The top-level NOTICE file has only 4 lines, you want me to move this copyright notice (line 21) to there?..

This doesn't make sense to me; this line here basically says "I wrote this file" (which is true); the same line in top-level NOTICE would say "I wrote the whole Apache Thrift" (which is blatantly not true). So, a very bad refactoring, that would be.

Just above line 21, it says *Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements*. I did sign ASF CLA. So, despite me claiming copyright credit for the file, ASF is perfectly allowed to do everything that Apache-2.0 license allows.

This is the golden arrangement for all parties here, at least according to my understanding.

Would you please explain? The page you linked talks mostly about dependencies, and doesn't describe my use case at all (contributing a new file).

Reply:
All correct, but we usually don't have copyright notices (**line 21**) in single files.

Reply:
LICENSE, sorry.

Reply:
@Jens-G 
> we usually don't have copyright notices

You sure about this? I see plenty, on the order of thousands. https://github.com/apache/thrift/search?q=copyright&unscoped_q=copyright

And in Open Source overall, this is pretty normal situation.

Reply:
> You sure about this? I see plenty, on the order of thousands.

Not to take any sides here, but I think the github search may be misleading. If you search the individual files, and remove irrelevant matches, it may look different. Here is an unrepresentative naiive search that I just did:
```
#> grep -Iirn "copyright" *|grep -v "node_modules\|regarding copyright ownership\|LICENSE\|Apache-License\|jQuery\|Google\|doc/licenses\|AssemblyCopyright\|The Apache Software Foundation\|aclocal\|debian/copyright\|debian/README.md\|gradlew\|build/cmake"
```
This returns
```
bda-build-project.sh:7:# Copyright (c) BioDataAnalysis GmbH. All Rights Reserved.
compiler/cpp/src/thrift/generate/t_cl_generator.cc:2: * Copyright (c) 2008- Patrick Collison <patrick@collison.ie>
compiler/cpp/src/thrift/generate/t_cl_generator.cc:3: * Copyright (c) 2006- Facebook
compiler/cpp/src/thrift/generate/t_html_generator.h:5:  " * Copyright 2012 Twitter, Inc\n"                                                               \
compiler/cpp/tests/catch/catch.hpp:6: *  Copyright (c) 2012 Two Blue Cubes Ltd. All rights reserved.
contrib/fb303/acinclude.m4:1:dnl Copyright (C) 2009 Facebook
contrib/fb303/acinclude.m4:3:dnl are permitted in any medium without royalty provided the copyright
debian/changelog:54:  * update copyright file
doc/specs/thrift.tex:17:\documentclass[nocopyrightspace,blockstyle]{sigplanconf}
doc/specs/thrift.tex:27:% \copyrightyear{2007}
doc/specs/thrift.tex:28:% \copyrightdata{[to be supplied]}
lib/delphi/test/keywords/ReservedKeywords.dproj:95:						<VersionInfoKeys Name="LegalCopyright"/>
lib/py/compat/win32/stdint.h:4://  Copyright (c) 2006-2008 Alexander Chemeris
lib/py/compat/win32/stdint.h:9://   1. Redistributions of source code must retain the above copyright notice,
lib/py/compat/win32/stdint.h:12://   2. Redistributions in binary form must reproduce the above copyright
lib/cpp/src/thrift/windows/SocketPair.cpp:2: * Copyright 2007 by Nathan C. Myers <ncm@cantrip.org>; some rights reserved.
lib/cpp/src/thrift/windows/SocketPair.cpp:6: * this copyright notice must be retained, unchanged, in its entirety.  If
lib/php/src/ext/thrift_protocol/config.m4:1:dnl Copyright (C) 2009 Facebook
lib/php/src/ext/thrift_protocol/config.m4:3:dnl are permitted in any medium without royalty provided the copyright
lib/js/test/build/js/lib/Int64.js:2048://     Copyright (c) 2012 Robert Kieffer
lib/js/test/build/js/lib/Int64Util.js:2143://     Copyright (c) 2012 Robert Kieffer
lib/js/test/build/js/lib/JSONInt64.js:2959://     Copyright (c) 2012 Robert Kieffer
```
Of course I have no idea how to do this correctly, so very likely I'm ignoring too much...

Reply:
To sum up: I can move the copyright line to the top of the file, seems it's more common to have it as close to the top as possible.

But I don't find it appropriate to comply to requests to remove it. I wrote this file, and I can credit myself for that via the copyright line.

There's no better place for the line than this file itself.

The line does no harm to ASF, rather to the contrary; because Apache-2.0 license, and because *more contributors* means *stronger foundation*.

Reply:
Yeah, you are a clever guy.

Reply:
Lets not get too hung up on this. @ulidtko I'm not a core team member but I do have a slight concern about the copyright that may or may not help to progress with this question. 

> This doesn't make sense to me; this line here basically says "I wrote this file" (which is true); the same line in top-level NOTICE would say "I wrote the whole Apache Thrift" (which is blatantly not true). So, a very bad refactoring, that would be.

I've several times had the opposite problem to yours. I made a small (but I felt relevant) contribution to a larger code file from somebody else. If such a code file had an `authors` list or a `copyright` notice, this brought a dilemma with itself: Should I add myself to the header for my small change, or should I go unnoticed? Going unnoticed feels wrong, but having to maintain long authors lists for every code file equally so.

While I do not mean to impose anything, and I'm not a Thrift member, I can see good reason for a single list of authors that contributed to the project. For anyone curious, git blame easily allows to see this in more detail for individual files (and many tools like Visual Studio show this automatically on a line-by-line base).

Does that change anything for you?

Reply:
Let him have his copyright line. I'll add it to license myself.